File: workers/worker-go/registry.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"os"
	"runtime"

	"github.com/redis/go-redis/v9"
)

const (
	CAT_CLASSES     = "Character Classes & Escapes"
	CAT_ANCHORS     = "Anchors & Boundaries"
	CAT_QUANTIFIERS = "Quantifiers"
	CAT_GROUPS      = "Grouping & Backreferences"
	CAT_ADVANCED    = "Lookarounds & Advanced"
)

func registerEngines(ctx context.Context, client *redis.Client) error {
	workerVersion := os.Getenv("WORKER_VERSION")
	if workerVersion == "" {
		workerVersion = "Unknown"
	}
	workerReleaseDate := os.Getenv("WORKER_RELEASE_DATE")
	if workerReleaseDate == "" {
		workerReleaseDate = "Unreleased"
	}

	engineInfo := EngineInfo{
		EngineID:              "go_standard",
		EngineLanguageType:    "Go",
		EngineLanguageVersion: runtime.Version(),
		EngineRegexLib:        "regexp",
		EngineRegexLibVersion: runtime.Version(),
		EngineLabel:           "Go (regexp)",
		EngineCapabilities: EngineCapabilities{
			Flags:              supportedGoRegexFlags(),
			SupportsLookaround: false,
			SupportsBackrefs:   false,
		},
		EngineDocs: EngineDocs{
			Trivia: []string{
				"Based on the RE2 syntax and algorithms.",
				"Go's standard library (including regexp) is distributed under a BSD-3-Clause style Go license.",
				"Guarantees linear time O(n) execution, preventing ReDoS.",
				"Supports inline flags i, m, s, and U.",
				"Does not support backreferences or lookaround assertions.",
			},
			CheatSheetURL: "https://pkg.go.dev/regexp/syntax",
		},
		EngineCheatSheet: []CheatSheetCategory{
			{
				Category: CAT_CLASSES,
				Items: []CheatSheetItem{
					{Character: ".", Description: "Any character except newline unless the 's' flag is set"},
					{Character: "\\w", Description: "ASCII word character ([0-9A-Za-z_])"},
					{Character: "\\W", Description: "ASCII non-word character"},
					{Character: "\\d", Description: "ASCII digit ([0-9])"},
					{Character: "\\D", Description: "ASCII non-digit"},
					{Character: "\\s", Description: "ASCII whitespace ([\\t\\n\\f\\r ])"},
					{Character: "\\S", Description: "ASCII non-whitespace character"},
					{Character: "[a-z]", Description: "Character class (inclusive)"},
					{Character: "[^a-z]", Description: "Negated character class (exclusive)"},
					{Character: "\\p{Greek}", Description: "Unicode character class"},
					{Character: "\\P{Greek}", Description: "Negated Unicode character class"},
				},
			},
			{
				Category: CAT_ANCHORS,
				Items: []CheatSheetItem{
					{Character: "^", Description: "Start of text (or line if the 'm' flag is set)"},
					{Character: "$", Description: "End of text (or line if the 'm' flag is set)"},
					{Character: "\\A", Description: "Absolute start of text"},
					{Character: "\\z", Description: "Absolute end of text"},
					{Character: "\\b", Description: "ASCII word boundary"},
					{Character: "\\B", Description: "ASCII non-word boundary"},
				},
			},
			{
				Category: CAT_QUANTIFIERS,
				Items: []CheatSheetItem{
					{Character: "*", Description: "0 or more times (greedy unless the 'U' flag is set)"},
					{Character: "+", Description: "1 or more times (greedy unless the 'U' flag is set)"},
					{Character: "?", Description: "0 or 1 time (greedy unless the 'U' flag is set)"},
					{Character: "{m}", Description: "Exactly m times"},
					{Character: "{m,n}", Description: "Between m and n times (greedy unless the 'U' flag is set)"},
					{Character: "*?", Description: "0 or more times (lazy unless the 'U' flag is set)"},
					{Character: "+?", Description: "1 or more times (lazy unless the 'U' flag is set)"},
					{Character: "??", Description: "0 or 1 time (lazy unless the 'U' flag is set)"},
					{Character: "{m,n}?", Description: "Between m and n times (lazy unless the 'U' flag is set)"},
				},
			},
			{
				Category: CAT_GROUPS,
				Items: []CheatSheetItem{
					{Character: "(...)", Description: "Capturing group"},
					{Character: "(?:...)", Description: "Non-capturing group"},
					{Character: "x|y", Description: "Alternation (match x or y)"},
					{Character: "(?P<name>...)", Description: "Named capturing group"},
					{Character: "(?<name>...)", Description: "Named capturing group"},
				},
			},
			{
				Category: CAT_ADVANCED,
				Items: []CheatSheetItem{
					{Character: "(?i)", Description: "Inline flag: case-insensitive matching"},
					{Character: "(?m)", Description: "Inline flag: multiline mode; ^ and $ match line boundaries"},
					{Character: "(?s)", Description: "Inline flag: dotall mode; . matches newline"},
					{Character: "(?U)", Description: "Inline flag: ungreedy mode; swaps greedy and lazy quantifiers"},
					{Character: "(?i:...)", Description: "Scoped inline flag group"},
					{Character: "(?-i:...)", Description: "Scoped inline flag clearing group"},
				},
			},
		},
		EngineExamples: []EngineExample{
			{
				Regex: "(?P<IP>(?:25[0-5]|2[0-4]\\d|1\\d\\d|[1-9]?\\d)\\.(?:25[0-5]|2[0-4]\\d|1\\d\\d|[1-9]?\\d)\\.(?:25[0-5]|2[0-4]\\d|1\\d\\d|[1-9]?\\d)\\.(?:25[0-5]|2[0-4]\\d|1\\d\\d|[1-9]?\\d))(/(\\d{1,2}))?(?:-(?P<IP2>(?:25[0-5]|2[0-4]\\d|1\\d\\d|[1-9]?\\d)\\.(?:25[0-5]|2[0-4]\\d|1\\d\\d|[1-9]?\\d)\\.(?:25[0-5]|2[0-4]\\d|1\\d\\d|[1-9]?\\d)\\.(?:25[0-5]|2[0-4]\\d|1\\d\\d|[1-9]?\\d)))?(:(?P<port>\\d{1,5}))?",
				Text:  "This is an example to get IP:\n\n192.168.1.100\n192.168.1.100:8080\n127.0.0.1\n192.168.1.0/24\n192.168.1.1-192.168.1.255\n192.168.1.1-192.168.1.255:80\n192.168.1.0/24:80",
			},
		},
	}

	workerInfo := WorkerInfo{
		WorkerName:        "worker-go",
		WorkerVersion:     workerVersion,
		WorkerReleaseDate: workerReleaseDate,
		Engines:           []EngineInfo{engineInfo},
	}

	jsonBytes, err := json.Marshal(workerInfo)
	if err != nil {
		return err
	}

	err = client.HSet(ctx, "openregex:workers", workerInfo.WorkerName, string(jsonBytes)).Err()
	if err == nil {
		fmt.Printf("[Discovery] Registered '%s' with %d engines.\n", workerInfo.WorkerName, len(workerInfo.Engines))
	}
	return err
}

File: workers/worker-php/src/Registry.php
<?php

namespace OpenRegex\Worker;

class Registry {
    private static function flagMetadata(string $name): array {
        return match ($name) {
            'i' => ['Case-insensitive matching.', 'Basic'],
            'm' => ['Multiline mode. Makes ^ and $ work per line.', 'Basic'],
            's' => ['DotAll mode. Makes . match newline.', 'Basic'],
            'x' => ['Extended/free-spacing pattern mode.', 'Basic'],
            'A' => ['Force pattern to be anchored at the start of the subject.', 'Advance'],
            'D' => ['$ matches only at the true end of the subject.', 'Advance'],
            'S' => ['Compatibility study modifier; ignored in modern PHP/PCRE2.', 'Unique'],
            'U' => ['Ungreedy mode; quantifiers are lazy by default.', 'Advance'],
            'X' => ['Extra syntax checking for unknown escaped letters.', 'Unique'],
            'J' => ['Allow duplicate named capturing groups.', 'Unique'],
            'u' => ['Treat subject and pattern as UTF-8.', 'Basic'],
            'n' => ['No auto-capture mode (PHP 8.2+).', 'Advance'],
            'r' => ['Restrict caseless ASCII/non-ASCII boundary folds (PHP 8.4+).', 'Unique'],
            default => ["Flag ({$name})", 'Basic'],
        };
    }

    private static function buildEngineFlags(array $names): array {
        $flags = [];
        foreach ($names as $name) {
            [$description, $group] = self::flagMetadata($name);
            $flags[] = [
                'name' => $name,
                'description' => $description,
                'group' => $group,
            ];
        }
        return $flags;
    }

    public static function registerEngines($client) {
        $workerVersion = getenv('WORKER_VERSION') ?: '1.0.0';
        $releaseDate = getenv('WORKER_RELEASE_DATE') ?: 'Unreleased';

        $pcreFlagNames = ['i', 'm', 's', 'x', 'A', 'D', 'S', 'U', 'X', 'J', 'u'];

        if (defined('PHP_VERSION_ID') && PHP_VERSION_ID >= 80200) {
            $pcreFlagNames[] = 'n';
        }

        if (defined('PHP_VERSION_ID') && PHP_VERSION_ID >= 80400) {
            $pcreFlagNames[] = 'r';
        }

        $workerInfo = [
            'worker_name' => 'worker-php',
            'worker_version' => $workerVersion,
            'worker_release_date' => $releaseDate,
            'engines' => [
                [
                    'engine_id' => 'php_pcre',
                    'engine_language_type' => 'PHP',
                    'engine_language_version' => phpversion(),
                    'engine_regex_lib' => 'PCRE',
                    'engine_regex_lib_version' => PCRE_VERSION,
                    'engine_label' => 'PHP (PCRE)',
                    'engine_capabilities' => [
                        'flags' => self::buildEngineFlags($pcreFlagNames),
                        'supports_lookaround' => true,
                        'supports_backrefs' => true
                    ],
                    'engine_docs' => [
                        'trivia' => [
                            "Powered by the highly compatible Perl Compatible Regular Expressions (PCRE/PCRE2) C library.",
                            "PCRE2 is distributed under BSD-3-Clause licensing with the PCRE2 exception.",
                            "The standard preg_* functions utilize an NFA backtracking engine.",
                            "Susceptible to ReDoS if patterns are poorly optimized; however, execution is protected internally via pcre.backtrack_limit.",
                            "The 'u' (PCRE_UTF8) modifier is automatically appended by OpenRegex to safely navigate string payloads."
                        ],
                        'cheat_sheet_url' => 'https://www.php.net/manual/en/reference.pcre.pattern.syntax.php'
                    ],
                    'engine_cheat_sheet' => [
                        [
                            'category' => 'Character Classes & Escapes',
                            'items' => [
                                ['character' => '.', 'description' => "Any character except newline unless 's' flag is set"],
                                ['character' => '\w', 'description' => 'Word character; ASCII by default, Unicode-aware with UCP mode'],
                                ['character' => '\W', 'description' => 'Non-word character'],
                                ['character' => '\d', 'description' => 'Decimal digit; ASCII by default, Unicode-aware with UCP mode'],
                                ['character' => '\D', 'description' => 'Non-digit'],
                                ['character' => '\s', 'description' => 'Whitespace character; ASCII by default, Unicode-aware with UCP mode'],
                                ['character' => '\S', 'description' => 'Non-whitespace character'],
                                ['character' => '\h', 'description' => 'Horizontal whitespace'],
                                ['character' => '\v', 'description' => 'Vertical whitespace'],
                                ['character' => '\p{L}', 'description' => 'Unicode letter property'],
                                ['character' => '\P{L}', 'description' => 'Negated Unicode letter property'],
                                ['character' => '[a-z]', 'description' => 'Character class'],
                                ['character' => '[^a-z]', 'description' => 'Negated character class']
                            ]
                        ],
                        [
                            'category' => 'Anchors & Boundaries',
                            'items' => [
                                ['character' => '^', 'description' => "Start of subject, or start of line if 'm' flag is set"],
                                ['character' => '$', 'description' => "End of subject or before final newline, or end of line if 'm' flag is set"],
                                ['character' => '\A', 'description' => 'Absolute start of subject'],
                                ['character' => '\z', 'description' => 'Absolute end of subject'],
                                ['character' => '\Z', 'description' => 'End of subject or before final newline'],
                                ['character' => '\G', 'description' => 'First matching position in subject'],
                                ['character' => '\b', 'description' => 'Word boundary'],
                                ['character' => '\B', 'description' => 'Non-word boundary']
                            ]
                        ],
                        [
                            'category' => 'Quantifiers',
                            'items' => [
                                ['character' => '*', 'description' => '0 or more times, greedy'],
                                ['character' => '+', 'description' => '1 or more times, greedy'],
                                ['character' => '?', 'description' => '0 or 1 time, greedy'],
                                ['character' => '{m}', 'description' => 'Exactly m times'],
                                ['character' => '{m,n}', 'description' => 'Between m and n times, greedy'],
                                ['character' => '*?', 'description' => '0 or more times, lazy'],
                                ['character' => '+?', 'description' => '1 or more times, lazy'],
                                ['character' => '??', 'description' => '0 or 1 time, lazy'],
                                ['character' => '{m,n}?', 'description' => 'Between m and n times, lazy'],
                                ['character' => '*+', 'description' => '0 or more times, possessive'],
                                ['character' => '++', 'description' => '1 or more times, possessive'],
                                ['character' => '?+', 'description' => '0 or 1 time, possessive'],
                                ['character' => '{m,n}+', 'description' => 'Between m and n times, possessive']
                            ]
                        ],
                        [
                            'category' => 'Grouping & Backreferences',
                            'items' => [
                                ['character' => '(...)', 'description' => 'Capturing group'],
                                ['character' => '(?:...)', 'description' => 'Non-capturing group'],
                                ['character' => 'x|y', 'description' => 'Alternation (match x or y)'],
                                ['character' => '(?<name>...)', 'description' => 'Named capturing group'],
                                ['character' => "(?'name'...)", 'description' => 'Named capturing group'],
                                ['character' => '(?P<name>...)', 'description' => 'Named capturing group'],
                                ['character' => '\1', 'description' => 'Backreference to capture group 1'],
                                ['character' => '\k<name>', 'description' => 'Backreference to named group'],
                                ['character' => '(?&name)', 'description' => 'Subroutine call to a named group'],
                                ['character' => '(?R)', 'description' => 'Recursive subroutine call to the whole pattern']
                            ]
                        ],
                        [
                            'category' => 'Lookarounds & Advanced',
                            'items' => [
                                ['character' => '(?=...)', 'description' => 'Positive lookahead'],
                                ['character' => '(?!...)', 'description' => 'Negative lookahead'],
                                ['character' => '(?<=...)', 'description' => 'Positive lookbehind with fixed-length alternatives'],
                                ['character' => '(?<!...)', 'description' => 'Negative lookbehind with fixed-length alternatives'],
                                ['character' => '(?>...)', 'description' => 'Atomic group'],
                                ['character' => '(*SKIP)(*F)', 'description' => 'Backtracking-control verbs commonly used to skip alternatives'],
                                ['character' => '\K', 'description' => 'Reset the start of the reported match'],
                                ['character' => '(?(condition)yes|no)', 'description' => 'Conditional subpattern'],
                                ['character' => '(?i)', 'description' => 'Inline flag: Case-insensitive'],
                                ['character' => '(?m)', 'description' => 'Inline flag: Multiline'],
                                ['character' => '(?s)', 'description' => 'Inline flag: Dot matches newline'],
                                ['character' => '(?x)', 'description' => 'Inline flag: Extended / free-spacing'],
                                ['character' => '(?U)', 'description' => 'Inline flag: Ungreedy quantifiers'],
                                ['character' => '(?J)', 'description' => 'Inline flag: Allow duplicate named groups']
                            ]
                        ]
                    ],
                    'engine_examples' => [
                        [
                            'regex' => '(?<IP>(?:25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)\.(?:25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)\.(?:25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)\.(?:25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d))(/(\d{1,2}))?(?:-(?<IP2>(?:25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)\.(?:25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)\.(?:25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)\.(?:25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)))?(?::(?<port>\d{1,5}))?',
                            'text' => "This is an example to get IP:\n\n192.168.1.100\n192.168.1.100:8080\n127.0.0.1\n192.168.1.0/24\n192.168.1.1-192.168.1.255\n192.168.1.1-192.168.1.255:80\n192.168.1.0/24:80"
                        ]
                    ]
                ]
            ]
        ];

        $client->hset('openregex:workers', 'worker-php', json_encode($workerInfo));
        echo "[Worker] Registered 'worker-php' with 1 engine.\n";
    }
}
